File: server/lib/libdb.ml
open Core
open Lib
open Runtime
open Types.RuntimeT


let fns : Lib.shortfn list = [

  { pns = ["DB::insert"]
  ; ins = []
  ; p = [par "val" TObj; par "table" TDB]
  ; r = TObj
  ; d = "Insert `val` into `table`"
  ; f = InProcess
        (function
          | (state, [DObj value; DDB db]) ->
            let id = Db.insert state.dbs db value in
            DObj (Map.set value "id" (DID id))
          | (_, args) -> fail args)
  ; pr = None
  ; ps = false
  }
  ;

  { pns = ["DB::delete"]
  ; ins = []
  ; p = [par "value" TObj; par "table" TDB]
  ; r = TNull
  ; d = "Delete `value` from `table`"
  ; f = InProcess
        (function
          | (state, [DObj vals; DDB db]) ->
            Db.delete state.dbs db vals;
            DNull
          | (_, args) -> fail args)
  ; pr = None
  ; ps = false
  }
  ;


  { pns = ["DB::deleteAll"]
  ; ins = []
  ; p = [par "table" TDB]
  ; r = TNull
  ; d = "Delete everything from `table`"
  ; f = InProcess
        (function
          | (state, [DDB db]) ->
            Db.delete_all state.dbs db;
            DNull
          | (_, args) -> fail args)
  ; pr = None
  ; ps = false
  }
  ;



  { pns = ["DB::update"]
  ; ins = []
  ; p = [par "value" TObj; par "table" TDB]
  ; r = TNull
  ; d = "Update `table` value which has the same ID as `value`"
  ; f = InProcess
        (function
          | (state, [DObj vals; DDB db]) ->
            Db.update state.dbs db vals;
            DObj vals
          | (_, args) -> fail args)
  ; pr = None
  ; ps = false
  }
  ;

  { pns = ["DB::fetchBy"]
  ; ins = []
  ; p = [par "value" TAny; par "field" TStr; par "table" TDB]
  ; r = TList
  ; d = "Fetch the value in `table` whose field `field` is `value`"
  ; f = InProcess
        (function
          | (state, [value; DStr field; DDB db]) ->
            Db.fetch_by state.dbs db field value
          | (_, args) -> fail args)
  ; pr = None
  ; ps = true
  }
  ;

  { pns = ["DB::fetchOneBy"]
  ; ins = []
  ; p = [par "value" TAny; par "field" TStr; par "table" TDB]
  ; r = TAny
  ; d = "Fetch exactly one value in `table` whose field `field` is `value`"
  ; f = InProcess
        (function
          | (state, [value; DStr field; DDB db]) ->
            let result = Db.fetch_by state.dbs db field value in
            (match result with
             | DList (x :: xs) -> x
               (* TODO(ian): Maybe/Option *)
             | _ -> DNull)
          | (_, args) -> fail args)
  ; pr = None
  ; ps = true
  }
  ;

  { pns = ["DB::fetchAll"]
  ; ins = []
  ; p = [par "table" TDB]
  ; r = TList
  ; d = "Fetch all the values in `table`"
  ; f = InProcess
        (function
          | (state, [DDB db]) ->
            Db.fetch_all state.dbs db
          | (_, args) -> fail args)
  ; pr = None
  ; ps = true
  }
  ;

  { pns = ["DB::keys"]
  ; ins = []
  ; p = [par "table" TDB]
  ; r = TList
  ; d = "Fetch all the keys in `table`"
  ; f = InProcess
        (function
          | (_, [DDB db]) ->
            Db.cols_for db
            |> List.map ~f:(fun (k,v) -> DStr k)
            |> DList
          | (_, args) -> fail args)
  ; pr = None
  ; ps = true
  }
  ;

  { pns = ["DB::schema"]
  ; ins = []
  ; p = [par "table" TDB]
  ; r = TObj
  ; d = "Fetch all the values in `table`"
  ; f = InProcess
        (function
          | (_, [DDB db]) ->
            Db.cols_for db
            |> List.map ~f:(fun (k,v) -> (k, DStr (Dval.tipe_to_string v)))
            |> Dval.to_dobj
          | (_, args) -> fail args)
  ; pr = None
  ; ps = true
  }
  ;

]

File: fsharp-backend/src/BackendOnlyStdLib/LibDB.fs
/// (deprecated) StdLib functions for accessing and manipulating user datastores
module BackendOnlyStdLib.LibDB

open Prelude
open LibExecution.RuntimeTypes

let fn = FQFnName.stdlibFnName

let err (str : string) = Ply(Dval.errStr str)

let removedFunction = LibExecution.Errors.removedFunction
let incorrectArgs = LibExecution.Errors.incorrectArgs

let varA = TVariable "a"
// CLEANUP: remove the obj type and use varA instead
let obj = TDict varA
let dbType = TDB varA

let fns : List<BuiltInFn> =
  [ { name = fn "DB" "insert" 0
      parameters = [ Param.make "val" obj ""; Param.make "table" dbType "" ]
      returnType = obj
      description = "Insert `val` into `table`"
      fn =
        function
        | state, [ v; table ] -> removedFunction state "DB::insert"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "delete" 0
      parameters = [ Param.make "value" obj ""; Param.make "table" dbType "" ]
      returnType = TNull
      description = "Delete `value` from `table`"
      fn =
        function
        | state, [ v; table ] -> removedFunction state "DB::delete"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "deleteAll" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TNull
      description = "Delete everything from `table`"
      fn =
        function
        | state, [ table ] -> removedFunction state "DB::deleteAll"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "update" 0
      parameters = [ Param.make "value" obj ""; Param.make "table" dbType "" ]
      returnType = TNull
      description = "Update `table` value which has the same ID as `value`"
      fn =
        function
        | state, [ v; table ] -> removedFunction state "DB::update"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "fetchBy" 0
      parameters =
        [ Param.make "value" varA ""
          Param.make "field" TStr ""
          Param.make "table" dbType "" ]
      returnType = TList varA
      description = "Fetch all the values in `table` whose `field` is `value`"
      fn =
        function
        | state, [ v; field; table ] -> removedFunction state "DB::fetchBy"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "fetchOneBy" 0
      parameters =
        [ Param.make "value" varA ""
          Param.make "field" TStr ""
          Param.make "table" dbType "" ]
      returnType = varA
      description = "Fetch exactly one value in `table` whose `field` is `value`"
      fn =
        function
        | state, [ v; field; table ] -> removedFunction state "DB::fetchOneBy"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "fetchByMany" 0
      parameters = [ Param.make "spec" obj ""; Param.make "table" dbType "" ]
      returnType = TList varA
      description =
        "Fetch all the values from `table` which have the same fields and values that `spec` has"
      fn =
        function
        | state, [ spec; table ] -> removedFunction state "DB::fetchByMany"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "fetchOneByMany" 0
      parameters = [ Param.make "spec" obj ""; Param.make "table" dbType "" ]
      returnType = varA
      description =
        "Fetch exactly one value from `table`, which have the same fields and values that `spec` has"
      fn =
        function
        | state, [ spec; table ] -> removedFunction state "DB::fetchOneByMany"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "fetchAll" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TList varA
      description = "Fetch all the values in `table`"
      fn =
        function
        | state, [ table ] -> removedFunction state "DB::fetchAll"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "keys" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TList varA
      description = "Fetch all the keys in `table`"
      fn =
        function
        | state, [ table ] -> removedFunction state "DB::keys"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") }


    { name = fn "DB" "schema" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = obj
      description = "Fetch all the values in `table`"
      fn =
        function
        | state, [ table ] -> removedFunction state "DB::schema"
        | _ -> incorrectArgs ()
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = DeprecatedBecause("Old DB functions have been removed") } ]
